rust/crates/sdk/src/stream.rs

```diff
@@ -0,0 +1,286 @@
+    pub async fn new(config: &Config, feed_ids: Vec<ID>) -> Result<Stream, StreamError> {
+        let (report_sender, report_receiver) = mpsc::channel(100);
+        let (shutdown_sender, _) = broadcast::channel(1);
+
+        let stats = Arc::new(Stats {
+            accepted: AtomicUsize::new(0),
+            deduplicated: AtomicUsize::new(0),
+            partial_reconnects: AtomicUsize::new(0),
+            full_reconnects: AtomicUsize::new(0),
+            configured_connections: AtomicUsize::new(0),
+            active_connections: AtomicUsize::new(0),
+        });
+
+        let conn = connect(config, &feed_ids, stats.clone()).await?;
+
+        let water_mark = Arc::new(Mutex::new(HashMap::new()));
+
+        Ok(Stream {
+            config: config.clone(),
+            feed_ids,
+            conn: Some(conn),
+            report_sender,
+            report_receiver,
+            shutdown_sender,
+            stats,
+            water_mark,
+        })
+    }
+
+    /// Starts listening for reports on the Stream.
+    /// This method will spawn a new task for each WebSocket connection.
+    pub async fn listen(&mut self) -> Result<(), StreamError> {
+        let conn = self
+            .conn
+            .take()
+            .ok_or_else(|| StreamError::ConnectionError("No connection".into()))?;
+
+        match conn {
+            WebSocketConnection::Single(stream) => {
+                let report_sender = self.report_sender.clone();
+                let shutdown_receiver = self.shutdown_sender.subscribe();
+                let stats = self.stats.clone();
+                let water_mark = self.water_mark.clone();
+                let config = self.config.clone();
+                let feed_ids = self.feed_ids.clone();
+
+                tokio::spawn(run_stream(
+                    stream,
+                    report_sender,
+                    shutdown_receiver,
+                    stats,
+                    water_mark,
+                    config,
+                    feed_ids,
+                ));
+            }
+            WebSocketConnection::Multiple(streams) => {
+                for stream in streams {
+                    let report_sender = self.report_sender.clone();
+                    let shutdown_receiver = self.shutdown_sender.subscribe();
+                    let stats = self.stats.clone();
+                    let water_mark = self.water_mark.clone();
+                    let config = self.config.clone();
+                    let feed_ids = self.feed_ids.clone();
+
+                    tokio::spawn(run_stream(
+                        stream,
+                        report_sender,
+                        shutdown_receiver,
+                        stats,
+                        water_mark,
+                        config,
+                        feed_ids,
+                    ));
+                }
+            }
+        }
```

Comment:
Optional little trick to compress this a bit:

```suggestion
              let streams: &mut dyn Iterator<
            Item = tokio_tungstenite::WebSocketStream<MaybeTlsStream<TcpStream>>,
        > = match conn {
            WebSocketConnection::Single(stream) => &mut Some(stream).into_iter(),
            WebSocketConnection::Multiple(streams) => &mut streams.into_iter(),
        };

        for stream in streams {
            let report_sender = self.report_sender.clone();
            let shutdown_receiver = self.shutdown_sender.subscribe();
            let stats = self.stats.clone();
            let water_mark = self.water_mark.clone();
            let config = self.config.clone();
            let feed_ids = self.feed_ids.clone();

            tokio::spawn(run_stream(
                stream,
                report_sender,
                shutdown_receiver,
                stats,
                water_mark,
                config,
                feed_ids,
            ));
        }

```